content/posts/0.27-release.md

```diff
@@ -0,0 +1,224 @@
+between 43% and 76% of the execution time was spent formatting the error message
+for the attribute error. We replaced the call to `PyObject_GenericGetAttr` with
+the private API `_PyObject_GenericGetAttrWithDict` which basically behaves the
+same as `PyObject_GenericGetAttr` but if the object isn't found, instead of
+formatting an `AttributeError` it simply returns `NULL`. If there is any other
+sort of error, the error flag is still set. This avoids wasting time formatting
+an error message that we are planning to throw away.
+
+The second improvement we made was to optimize away temporary bound methods.
+Python code like
+```py
+a.f()
+```
+means first look up `a.f` and then call the result, which is translated into
+bytecode like:
+```
+LOAD a
+LOAD_ATTR f
+CALL
+```
+The function `a.f` receives `a` as the first `self` argument, so it needs to be
+a special bound method object that knows about the `self` argument. However, we
+only use this object once to call it and then we are going to throw it away. We
+can avoid allocating and destroying an object when we call a method by calling
+`type(a).f(a)`. The `LOAD_ATTR` opcode has a special argument that indicates
+that the next opcode is going to be `CALL` and in that case calls a method named
+`_PyObject_GetMethod` instead of the typical `PyObject_GetAttr` to perform the
+attribute lookup. We patched `_PyObject_GetMethod` to have special handling for
+`JsProxy` objects so that we could optimize away the temporary for JS objects too.
+
+### A potential future performance improvement
+
+In the case of lookup on a `JsProxy`, there are other bytecode patterns where we
+could optimize away temporaries. For instance, if `a` is a `JsProxy` for the
+JavaScript object `{ x: { y: 2 } }` then `a.x.y` creates a temporary `JsProxy`
+pointing to `{ y: 2 }` and then looks up `y` on the temporary and returns `2`.
+The bytecode looks like:
+```
+LOAD a
+LOAD_ATTR x
+LOAD_ATTR y
+```
+If we know that we were going to use the result of the first `LOAD_ATTR x`
+immediately, we could avoid creating the temporary. Unfortunately, unlike the
+method call example, native Python code doesn't create a temporary here so
+optimizing this would involve more invasive patches to the interpreter than
+removing the temporary bound methods.
+
+## Work in Progress / Roadmap
+
+For Pyodide 0.28, we hope to unvendor the package recipes from the main Pyodide
+repository. The package index will gain a separate release process from the
+Pyodide runtime itself. This should have several significant benefits:
+
+* It will reduce CI usage due to rebuilding packages less often.
+* People who add a package will not have to wait as long until their package is
+  included in our index.
+
+We are working on adding Emscripten back as a tier 3 target for Python. As a
+part of this work, we have recently upstreamed a large number of fixes to the
+Emscripten file system. This will make a lot of obscure h
```

Comment:
@hoodmane, was this sentence missed when you were writing it, by chance?

Reply:
I just pushed my local state as-is because I was not actively working on it and wanted to avoid conflicts.